sql/tests: TestRandomSyntaxSQLSmith failed

sql/tests.TestRandomSyntaxSQLSmith [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2937582&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=2937582&tab=artifacts#/) on master @ [445175daed0848429219755703febfb3b50a73e9](https://github.com/cockroachdb/cockroach/commits/445175daed0848429219755703febfb3b50a73e9):


```
        		SET CLUSTER SETTING sql.stats.histogram_collection.enabled = false;
        		SET CLUSTER SETTING sql.defaults.interleaved_tables.enabled = true;
        	CREATE TABLE table1 (col1_0 BOOL NULL, col1_1 GEOGRAPHY[] NOT NULL, col1_2 INTERVAL, col1_3 BIT(7) NOT NULL, col1_4 FLOAT4, col1_5 STRING NULL AS (lower(CAST(col1_0 AS STRING))) STORED, INDEX (col1_5, col1_2 ASC, col1_4), UNIQUE (col1_0 DESC, col1_4 ASC, col1_3 DESC), INDEX (col1_2, col1_3, col1_0 DESC) WHERE ((table1.col1_4 = '-Inf':::FLOAT8) OR (table1.col1_5 < e'\x00':::STRING)) OR (NOT table1.col1_0), INVERTED INDEX (col1_2 ASC, col1_4, col1_3 DESC, col1_5 ASC, col1_0 DESC, col1_1) WHERE ((NOT table1.col1_0) AND (table1.col1_4 = 1.401298464324817e-45:::FLOAT8)) OR (table1.col1_5 < 'X':::STRING), FAMILY (col1_0), FAMILY (col1_5, col1_3), FAMILY (col1_1), FAMILY (col1_2, col1_4));
        CREATE TABLE table2 (col2_0 GEOGRAPHY, col2_1 FLOAT8, col2_2 VARCHAR NULL, col2_3 "char" NOT NULL, col2_4 BOOL NOT NULL, col2_5 BIT(22), col2_6 GEOMETRY, col2_7 STRING AS (lower(CAST(col2_0 AS STRING))) STORED, col2_8 STRING AS (lower(CAST(col2_6 AS STRING))) STORED, col2_9 STRING AS (lower(CAST(col2_6 AS STRING))) VIRTUAL, col2_10 STRING NOT NULL AS (lower(col2_3)) STORED, col2_11 STRING AS (lower(CAST(col2_5 AS STRING))) STORED, col2_12 STRING AS (lower(CAST(col2_6 AS STRING))) VIRTUAL, UNIQUE (col2_9 ASC), INDEX (col2_9) WHERE (((((((((table2.col2_7 > e'\x00':::STRING) AND (table2.col2_2 <= '':::STRING)) AND (table2.col2_3 > e'\x00':::STRING)) OR (table2.col2_1 <= 'NaN':::FLOAT8)) OR (table2.col2_12 > '':::STRING)) OR table2.col2_4) AND (table2.col2_11 = '':::STRING)) OR (table2.col2_9 = e'\'':::STRING)) OR (table2.col2_8 > 'X':::STRING)) OR (table2.col2_10 < e'\x00':::STRING));
        CREATE TABLE table3 (col3_0 DATE NOT NULL, col3_1 CHAR NULL, col3_2 FLOAT4 NOT NULL, col3_3 UUID NULL, col3_4 NAME NOT NULL, col3_5 UUID NULL, col3_6 DECIMAL NOT NULL, col3_7 DATE NOT NULL, col3_8 DATE NOT NULL, PRIMARY KEY (col3_4, col3_8 ASC, col3_2 ASC, col3_6), col3_9 STRING NOT NULL AS (lower(CAST(col3_8 AS STRING))) STORED, col3_10 STRING NULL AS (lower(CAST(col3_5 AS STRING))) VIRTUAL, col3_11 DECIMAL NOT NULL AS (col3_6 + 5.131527692005656151E+28:::DECIMAL) STORED, UNIQUE (col3_9 ASC, col3_8 ASC) STORING (col3_0, col3_1, col3_3, col3_5, col3_7), UNIQUE (col3_1, col3_10 DESC, col3_7, col3_11 DESC, col3_4, col3_2, col3_6 DESC, col3_3, col3_8 ASC, col3_0 ASC, col3_9 ASC, col3_5 ASC), UNIQUE (col3_9, col3_0 ASC, col3_11 DESC, col3_10 ASC, col3_7, col3_1 DESC, col3_3 DESC, col3_6 ASC, col3_2 DESC), UNIQUE (col3_10 DESC, col3_0, col3_2 ASC, col3_9 ASC, col3_1 DESC, col3_7, col3_8, col3_6 DESC) STORING (col3_3, col3_5) WHERE ((((table3.col3_4 >= e'\'':::STRING) AND (table3.col3_11 > 1:::DECIMAL)) OR (table3.col3_6 >= 0:::DECIMAL)) OR (table3.col3_9 < '':::STRING)) OR (table3.col3_0 = '5874897-12-31':::DATE), INDEX (col3_1 ASC, col3_5, col3_9, col3_7, col3_0 ASC, col3_10 ASC, col3_3 ASC, col3_6, col3_4 ASC, col3_11 DESC, col3_2 DESC, col3_8 DESC) WHERE ((table3.col3_4 != '':::STRING) AND (table3.col3_7 < 'infinity':::DATE)) OR (table3.col3_6 = 0:::DECIMAL), INDEX (col3_8), UNIQUE (col3_1, col3_2, col3_8 ASC) WHERE table3.col3_2 >= (-1.0):::FLOAT8, INDEX (col3_0 ASC, col3_8, col3_7 ASC, col3_3, col3_9 DESC, col3_6, col3_5 ASC, col3_2, col3_4 ASC), INDEX (col3_6, col3_7 ASC, col3_5, col3_9, col3_3 ASC, col3_11 DESC, col3_10 ASC, col3_4 ASC, col3_1 ASC, col3_0 DESC, col3_2, col3_8 DESC) WHERE ((((table3.col3_4 != 'X':::STRING) AND (table3.col3_11 != 0:::DECIMAL)) AND (table3.col3_9 > e'\U00002603':::STRING)) OR (table3.col3_6 < '-Infinity':::DECIMAL)) AND (table3.col3_1 > e'\'':::STRING), FAMILY (col3_4), FAMILY (col3_8), FAMILY (col3_0, col3_1), FAMILY (col3_9), FAMILY (col3_2, col3_3), FAMILY (col3_6, col3_5, col3_7), FAMILY (col3_11));
        CREATE TABLE table4 (col4_0 GEOMETRY NOT NULL, col4_1 UUID[] NOT NULL, col4_2 DECIMAL NULL, col4_3 BOX2D NULL, col4_4 INT4 NULL, col4_5 DECIMAL NULL, col4_6 BOOL NOT NULL, col4_7 UUID NOT NULL, col4_8 REGPROC NULL, col4_9 REGCLASS, FAMILY (col4_9), FAMILY (col4_1), FAMILY (col4_3, col4_4, col4_5), FAMILY (col4_8, col4_6, col4_7, col4_0), FAMILY (col4_2));
        ALTER TABLE table1 INJECT STATISTICS e'[{"columns": ["col1_3"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 1438310005463877121, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 6519804461879664758}, {"columns": ["col1_4"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 4286084811104179795, "histo_col_type": "", "name": "__auto__", "null_count": 4567549032082631878, "row_count": 6519804461879664758}, {"columns": ["col1_5"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 3924342093668869320, "histo_buckets": [{"distinct_range": 0, "num_eq": 7000000000, "num_range": 0, "upper_bound": "\'"}, {"distinct_range": 0, "num_eq": 6114217279730797422, "num_range": 0, "upper_bound": "S(: )"}, {"distinct_range": 1549331779992475400, "num_eq": 3000000000, "num_range": 2561704201393771124, "upper_bound": "o~BY\\u0011\\u0001dx"}, {"distinct_range": 0, "num_eq": 30, "num_range": 0, "upper_bound": "y\\u0005c\\u0012cv\\u0004?M"}], "histo_col_type": "STRING", "name": "__auto__", "null_count": 3267620060525343433, "row_count": 6519804461879664758}, {"columns": ["col1_0"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 2268267903159141995, "histo_col_type": "BOOL", "name": "__auto__", "null_count": 558099816391872289, "row_count": 6519804461879664758}, {"columns": ["col1_1"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 1445192768762615382, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 6519804461879664758}, {"columns": ["col1_2"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 504677786827532923, "histo_buckets": [{"distinct_range": 0, "num_eq": 4209283920383281971, "num_range": 0, "upper_bound": "00:00:00"}, {"distinct_range": 0, "num_eq": 4264082944367446375, "num_range": 40000000, "upper_bound": "20 years 5 mons 563 days 22:58:10.444039"}, {"distinct_range": 30000000000, "num_eq": 3057261944973277047, "num_range": 30000000000, "upper_bound": "29 years 7 mons 947 days 16:49:41.571954"}, {"distinct_range": 0, "num_eq": 2631156809365080, "num_range": 1193538183244864306, "upper_bound": "38 years 4 mons 687 days 13:31:55.538653"}, {"distinct_range": 847163717742168300, "num_eq": 100000, "num_range": 1498429760327390591, "upper_bound": "41 years 10 mons 567 days 19:01:28.016766"}, {"distinct_range": 7695667539006467000, "num_eq": 100000, "num_range": 7695667539006467152, "upper_bound": "76 years 11 mons 726 days 22:30:03.007021"}, {"distinct_range": 0, "num_eq": 70000, "num_range": 8248995285534126318, "upper_bound": "77 years 3 mons 712 days 21:45:00.285976"}], "histo_col_type": "INTERVAL", "name": "__auto__", "null_count": 3885211005022210925, "row_count": 6519804461879664758}]':::JSONB;
        ALTER TABLE table2 INJECT STATISTICS e'[{"columns": ["col2_8"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 1688797457264484031, "histo_col_type": "", "name": "__auto__", "null_count": 4733644132164847081, "row_count": 6359371409903808125}, {"columns": ["col2_9"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 1177978033996963761, "histo_buckets": [{"distinct_range": 0, "num_eq": 600000000, "num_range": 0, "upper_bound": "\\u000c\\r\\u0014"}, {"distinct_range": 124029.74190429627, "num_eq": 30000000000, "num_range": 600000, "upper_bound": "\\u0010\\u00163.\\\\"}, {"distinct_range": 576.0283119496585, "num_eq": 5047740423037260757, "num_range": 5000, "upper_bound": "P>\\u0016\\u000ci\\\\\\u001d"}, {"distinct_range": 5408513338.016547, "num_eq": 100, "num_range": 10000000000, "upper_bound": "e"}, {"distinct_range": 3339489740936204000, "num_eq": 1811296971261832797, "num_range": 3339489740936203995, "upper_bound": "\U00002603"}], "histo_col_type": "STRING", "name": "__auto__", "null_count": 5483945034944615069, "row_count": 6359371409903808125}, {"columns": ["col2_5"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 4975515030414442202, "histo_col_type": "", "name": "__auto__", "null_count": 1288738071146887606, "row_count": 6359371409903808125}, {"columns": ["col2_6"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 5367428984424094345, "histo_col_type": "", "name": "__auto__", "null_count": 6178612714493878435, "row_count": 6359371409903808125}, {"columns": ["col2_7"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 3299881388045575677, "histo_col_type": "", "name": "__auto__", "null_count": 3719395029708790330, "row_count": 6359371409903808125}, {"columns": ["col2_0"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 1299681406484146327, "histo_col_type": "", "name": "__auto__", "null_count": 5669804889877757018, "row_count": 6359371409903808125}, {"columns": ["col2_1"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 1415647047254010867, "histo_col_type": "", "name": "__auto__", "null_count": 1790085091421075503, "row_count": 6359371409903808125}, {"columns": ["col2_2"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 3893720578183413627, "histo_col_type": "", "name": "__auto__", "null_count": 2455325823473704798, "row_count": 6359371409903808125}, {"columns": ["col2_3"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 3182765729432938880, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 6359371409903808125}, {"columns": ["col2_4"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 2602440718826433420, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 6359371409903808125}, {"columns": ["col2_10"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 3464113067522824044, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 6359371409903808125}, {"columns": ["col2_11"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 2440183147520938300, "histo_col_type": "", "name": "__auto__", "null_count": 4818560967003969391, "row_count": 6359371409903808125}, {"columns": ["col2_12"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 5284110397474033895, "histo_col_type": "", "name": "__auto__", "null_count": 4345921643779760542, "row_count": 6359371409903808125}]':::JSONB;
        ALTER TABLE table3 INJECT STATISTICS e'[{"columns": ["col3_6"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 1326225563027175029, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 1633053750447279447}, {"columns": ["col3_7"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 751942714549818761, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 1633053750447279447}, {"columns": ["col3_9"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 1388576773095937126, "histo_col_type": "STRING", "name": "__auto__", "null_count": 0, "row_count": 1633053750447279447}, {"columns": ["col3_10"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 1032415506344855869, "histo_col_type": "", "name": "__auto__", "null_count": 1268390833040457708, "row_count": 1633053750447279447}, {"columns": ["col3_0"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 777378313446563025, "histo_buckets": [{"distinct_range": 0, "num_eq": 1866376371770528104, "num_range": 0, "upper_bound": "1980-11-02"}, {"distinct_range": 7000, "num_eq": 1415770027770856310, "num_range": 7000, "upper_bound": "1983-12-01"}, {"distinct_range": 8910689835962711000, "num_eq": 20000, "num_range": 8910689835962710613, "upper_bound": "1985-01-01"}, {"distinct_range": 0, "num_eq": 70, "num_range": 100, "upper_bound": "1991-09-03"}, {"distinct_range": 0, "num_eq": 29878787913807470, "num_range": 694504577591496970, "upper_bound": "infinity"}], "histo_col_type": "DATE", "name": "__auto__", "null_count": 0, "row_count": 1633053750447279447}, {"columns": ["col3_1"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 1389499434770858346, "histo_buckets": [{"distinct_range": 0, "num_eq": 700000000, "num_range": 0, "upper_bound": "\\u0008"}, {"distinct_range": 4537896425.193343, "num_eq": 40000, "num_range": 8000000000, "upper_bound": "\\u0010"}, {"distinct_range": 661.4590591521817, "num_eq": 1000, "num_range": 700, "upper_bound": "\\u0018"}, {"distinct_range": 239.02494437399767, "num_eq": 6285621269083731521, "num_range": 1000, "upper_bound": "J"}, {"distinct_range": 0, "num_eq": 70000000000, "num_range": 6117766630918265424, "upper_bound": "n"}, {"distinct_range": 1000000000, "num_eq": 20000, "num_range": 1000000000, "upper_bound": "r"}, {"distinct_range": 3864.235027900915, "num_eq": 10000000, "num_range": 10000, "upper_bound": "\x7f"}], "histo_col_type": "CHAR", "name": "__auto__", "null_count": 1476703706590941572, "row_count": 1633053750447279447}, {"columns": ["col3_2"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 634313018724526083, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 1633053750447279447}, {"columns": ["col3_4"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 292020387134696428, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 1633053750447279447}, {"columns": ["col3_3"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 158159692498467617, "histo_col_type": "", "name": "__auto__", "null_count": 1229945961430235869, "row_count": 1633053750447279447}, {"columns": ["col3_5"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 162281805987333348, "histo_col_type": "", "name": "__auto__", "null_count": 834473399962435603, "row_count": 1633053750447279447}, {"columns": ["col3_8"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 1631903078221195200, "histo_buckets": [{"distinct_range": 0, "num_eq": 50000000, "num_range": 0, "upper_bound": "-infinity"}, {"distinct_range": 1838.2432225885727, "num_eq": 7066595047731010523, "num_range": 6000, "upper_bound": "1970-03-26"}, {"distinct_range": 0, "num_eq": 100000, "num_range": 0, "upper_bound": "1974-02-04"}, {"distinct_range": 0, "num_eq": 20000000, "num_range": 4022678466324979112, "upper_bound": "1986-08-20"}, {"distinct_range": 0, "num_eq": 2439675130958297270, "num_range": 5193663102717287822, "upper_bound": "1990-04-21"}, {"distinct_range": 1151337.0969159, "num_eq": 8749123426477954923, "num_range": 2000000, "upper_bound": "1993-09-10"}, {"distinct_range": 2835354574204394500, "num_eq": 70000000, "num_range": 4184665136148430173, "upper_bound": "1995-01-22"}, {"distinct_range": 1000000, "num_eq": 556412071463649610, "num_range": 1000000, "upper_bound": "1996-11-17"}], "histo_col_type": "DATE", "name": "__auto__", "null_count": 0, "row_count": 1633053750447279447}, {"columns": ["col3_11"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 991628977238660955, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 1633053750447279447}]':::JSONB;
        ALTER TABLE table4 INJECT STATISTICS '[{"columns": ["col4_9"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 15430799768, "histo_col_type": "", "name": "__auto__", "null_count": 41539338175, "row_count": 100000000000}, {"columns": ["col4_0"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 25871017658, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 100000000000}, {"columns": ["col4_5"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 18002181494, "histo_col_type": "", "name": "__auto__", "null_count": 53201241925, "row_count": 100000000000}, {"columns": ["col4_6"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 62681921555, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 100000000000}, {"columns": ["col4_7"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 11705206453, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 100000000000}, {"columns": ["col4_8"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 10703745294, "histo_col_type": "", "name": "__auto__", "null_count": 34303412676, "row_count": 100000000000}, {"columns": ["col4_1"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 80210521347, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 100000000000}, {"columns": ["col4_2"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 90518373727, "histo_col_type": "", "name": "__auto__", "null_count": 76386237622, "row_count": 100000000000}, {"columns": ["col4_3"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 58177975161, "histo_col_type": "", "name": "__auto__", "null_count": 80689051680, "row_count": 100000000000}, {"columns": ["col4_4"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 39097187982, "histo_col_type": "", "name": "__auto__", "null_count": 9566575652, "row_count": 100000000000}]':::JSONB;
        CREATE TYPE rand_typ_0 AS ENUM ('a', 'lp');
        CREATE TYPE rand_typ_1 AS ENUM ('q', 'efichu', 'sm', 'fqjagu');
        CREATE TYPE rand_typ_2 AS ENUM ('xs', 'dhkk');
        CREATE TYPE rand_typ_3 AS ENUM ('dmvf', 'wswtk', 'zm', 'dtcv', 'ljyvv');
        ;
    rsg_test.go:577: 
        SET CLUSTER SETTING sql.defaults.drop_enum_value.enabled = true;
        SET enable_drop_enum_value = true;
        
        CREATE TYPE greeting AS ENUM ('hello', 'howdy', 'hi', 'good day', 'morning');
        CREATE TABLE IF NOT EXISTS seed AS
        	SELECT
        		g::INT2 AS _int2,
        		g::INT4 AS _int4,
        		g::INT8 AS _int8,
        		g::FLOAT4 AS _float4,
        		g::FLOAT8 AS _float8,
        		'2001-01-01'::DATE + g AS _date,
        		'2001-01-01'::TIMESTAMP + g * '1 day'::INTERVAL AS _timestamp,
        		'2001-01-01'::TIMESTAMPTZ + g * '1 day'::INTERVAL AS _timestamptz,
        		g * '1 day'::INTERVAL AS _interval,
        		g % 2 = 1 AS _bool,
        		g::DECIMAL AS _decimal,
        		g::STRING AS _string,
        		g::STRING::BYTES AS _bytes,
        		substring('00000000-0000-0000-0000-' || g::STRING || '00000000000', 1, 36)::UUID AS _uuid,
        		'0.0.0.0'::INET + g AS _inet,
        		g::STRING::JSONB AS _jsonb,
        		enum_range('hello'::greeting)[g] as _enum
        	FROM
        		generate_series(1, 5) AS g;
        
        INSERT INTO seed DEFAULT VALUES;
        CREATE INDEX on seed (_int8, _float8, _date);
        CREATE INVERTED INDEX on seed (_jsonb);
        ;
    rsg_test.go:579: 
    rsg_test.go:580: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestRandomSyntaxSQLSmith460792454
--- FAIL: TestRandomSyntaxSQLSmith (300.69s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestRandomSyntaxSQLSmith PKG=./pkg/sql/tests TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #64010 sql/tests: TestRandomSyntaxSQLSmith failed [C-test-failure O-robot branch-release-21.1]
</p>
</details>
<details><summary>Internal log</summary>
<p>

```
mjibson marked as alumn{us/a}; resolving to rafiss instead
```
</p>
</details>

/cc @cockroachdb/sql-queries @rafiss
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxSQLSmith.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/logictest/testdata/logic_test/sqlsmith
+++ b/pkg/sql/logictest/testdata/logic_test/sqlsmith
@@ -157,3 +157,32 @@ query T
 SELECT ARRAY[1][NULL]
 ----
 NULL
+
+# Regression: #64399 (panic due to optimizer ordering bug)
+statement ok
+CREATE TABLE ab (
+  a INT,
+  b INT AS (a % 2) STORED,
+  INDEX (b)
+);
+CREATE TABLE cd (
+  c INT,
+  d INT AS (c % 2) VIRTUAL
+);
+
+statement ok
+SELECT NULL
+FROM ab AS ab1
+  JOIN cd AS cd1
+    JOIN cd AS cd2 ON cd1.c = cd2.c
+    JOIN ab AS ab2 ON
+      cd2.c = ab2.a
+      AND cd2.c = ab2.crdb_internal_mvcc_timestamp
+      AND cd1.c = ab2.a
+      AND cd1.c = ab2.b
+      AND cd1.d = ab2.b
+    ON ab1.b = cd1.d
+  JOIN cd AS cd3 ON
+    ab1.b = cd3.c
+    AND cd2.c = cd3.d
+    AND cd1.d = cd3.c

--- a/pkg/sql/opt/ordering/project.go
+++ b/pkg/sql/opt/ordering/project.go
@@ -22,18 +22,15 @@ func projectCanProvideOrdering(expr memo.RelExpr, required *physical.OrderingCho
 	proj := expr.(*memo.ProjectExpr)
 	inputCols := proj.Input.Relational().OutputCols
 
-	if required.CanProjectCols(inputCols) {
-		return true
-	}
-
-	// We may be able to "remap" columns using the internal FD set.
-	if fdSet := proj.InternalFDs(); required.CanSimplify(fdSet) {
-		simplified := required.Copy()
+	// Use a simplified ordering if it exists. This must be kept consistent with
+	// projectBuildChildReqOrdering, which always simplifies the ordering if
+	// possible.
+	simplified := *required
+	if fdSet := proj.InternalFDs(); simplified.CanSimplify(fdSet) {
+		simplified = required.Copy()
 		simplified.Simplify(fdSet)
-		return simplified.CanProjectCols(inputCols)
 	}
-
-	return false
+	return simplified.CanProjectCols(inputCols)
 }
 
 func projectBuildChildReqOrdering(

--- a/pkg/sql/opt/ordering/project_test.go
+++ b/pkg/sql/opt/ordering/project_test.go
@@ -39,7 +39,7 @@ func TestProject(t *testing.T) {
 
 	input := &testexpr.Instance{
 		Rel: &props.Relational{
-			OutputCols: opt.MakeColSet(1, 2, 3, 4),
+			OutputCols: opt.MakeColSet(1, 2, 3, 4, 6),
 			FuncDeps:   fds,
 		},
 	}
@@ -71,6 +71,14 @@ func TestProject(t *testing.T) {
 		},
 		{
 			req: "+5",
+			exp: "no",
+		},
+		{
+			// Regression test for #64399. projectCanProvideOrdering should not
+			// return true when the columns remaining in the ordering after
+			// simplification cannot be provided. This causes
+			// projectBuildChildReqOrdering to panic.
+			req: "+(5|6)",
 			exp: "no",
 		},
 	}

